File: LeetCode/56.Merge_Intervals/56.Merge_Intervals.py
#!/usr/bin/env python3


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # sort intervals
        # intervals[i][1] >= intervals[i+1][0], then merge

        intervals.sort(key=lambda x: x[0])
        res = []
        for e in intervals:
            if res and e[0] <= res[-1][1]:
                last = res[-1]
                res[-1] = [last[0], max(last[1], e[1])]
            else:
                res.append(e)

        return res

File: LeetCode/56.Merge_Intervals/56.Merge_Intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        PriorityQueue<int[]> heap = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        for (int[] interval : intervals) {
            heap.offer(interval);
        }

        Deque<int[]> q = new ArrayDeque<>();

        while (!heap.isEmpty()) {
            int[] first = heap.poll();
            if (q.isEmpty() || first[0] > q.peekLast()[1]) {
                q.offerLast(first);
            } else {
                int[] last = q.pollLast();
                last[0] = last[0];
                last[1] = Math.max(last[1], first[1]);
                heap.offer(last);
            }
        }

        int[][] res = new int[q.size()][2];
        for (int i = 0; i < res.length; i++) {
            res[i] = q.pollFirst();
        }

        return res;
    }
}